lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,260 @@
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = AggregationValue.new(name, COUNT, 0, tags, no_prefix)
+          end
+          aggregation_state[key].value += value
+        end
+      end
+
+      def aggregate_timing(name, value, tags: [], no_prefix: false, type: DISTRIBUTION)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).d(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, type)
+
+        mutex.synchronize do
+          if aggregation_state.key?(key) && aggregation_state[key].value.size + 1 >= @max_values
+            do_flush
+          end
+          if aggregation_state.key?(key)
+            aggregation_state[key].value << value
+          else
+            aggregation_state[key] = AggregationValue.new(name, type, [value], tags, no_prefix)
+          end
+        end
+      end
+
+      def gauge(name, value, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = AggregationValue.new(name, GAUGE, value, tags, no_prefix)
+          end
+          # keep the last value
+          aggregation_state[key].value = value
+        end
+      end
+
+      def flush
+        mutex.synchronize { do_flush }
+      end
+
+      private
+
+      def do_flush
+        @aggregation_state.each do |_key, agg|
+          case agg.type
+          when COUNT
+            @sink << datagram_builder(no_prefix: agg.no_prefix).c(
+              agg.name,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              agg.tags,
+            )
+          when DISTRIBUTION, MEASURE, HISTOGRAM
+            @sink << datagram_builder(no_prefix: agg.no_prefix).distribution_value_packed(
+              agg.name,
+              agg.type.to_s,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              agg.tags,
+            )
+          when GAUGE
+            @sink << datagram_builder(no_prefix: agg.no_prefix).g(
+              agg.name,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              agg.tags,
+            )
+          else
+            StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{agg.type}" }
+          end
+        end
+        @aggregation_state.clear
+      end
+
+      attr_reader :mutex, :aggregation_state, :sink
+
+      # @return Array<String>
+      def tags_sorted(tags)
+        return [].freeze if tags.nil? || tags.empty?
```

Comment:
If you're going to return a frozen aaray, you can constantize it to avoid an allocation.